File: orchestration/utils/logger.py
"""
@file logger.py
@module orchestration/utils/logger

Syncs the Python logging level with the backend Config table.
"""

import logging

from orchestration.utils.config import get_config

logger = logging.getLogger(__name__)

_last_level: str | None = None

LOG_LEVEL_MAP: dict[str, str] = {
    'debug': 'DEBUG',
    'info': 'INFO',
    'warn': 'WARNING',
    'warning': 'WARNING',
    'error': 'ERROR',
    'fatal': 'CRITICAL',
    'critical': 'CRITICAL',
    'trace': 'DEBUG',
}


def sync_log_level() -> None:
    """Read logging.level from Config cache and update the root logger."""
    global _last_level
    raw = get_config("logging.level", "INFO")
    level_name = LOG_LEVEL_MAP.get(str(raw).lower(), str(raw).upper())
    if level_name not in ("DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"):
        level_name = "INFO"

    if _last_level is not None and _last_level != level_name:
        print(f"🔄 [Config] Log level changed: {_last_level} → {level_name}")

    _last_level = level_name
    logging.getLogger().setLevel(getattr(logging, level_name, logging.INFO))

File: frontend/src/utils/logger.ts
/**
 * @file logger.ts
 * @module utils/logger
 *
 * Frontend logger that respects the logging.level config value.
 */

type LogFn = (msg: string, data?: unknown) => void;

let currentLevel = 'info';

const LEVEL_ORDER = ['debug', 'info', 'warn', 'error', 'fatal'] as const;

function shouldLog(level: string): boolean {
  const idx = LEVEL_ORDER.indexOf(level as typeof LEVEL_ORDER[number]);
  const cur = LEVEL_ORDER.indexOf(currentLevel as typeof LEVEL_ORDER[number]);
  return idx >= cur;
}

function make(level: string): LogFn {
  return (msg: string, data?: unknown) => {
    if (!shouldLog(level)) return;
    switch (level) {
      case 'debug': console.debug(msg, data ?? ''); break;
      case 'info':  console.info(msg, data ?? ''); break;
      case 'warn':  console.warn(msg, data ?? ''); break;
      case 'error': console.error(msg, data ?? ''); break;
      case 'fatal': console.error(`[FATAL] ${msg}`, data ?? ''); break;
    }
  };
}

export const logger = {
  debug: make('debug'),
  info: make('info'),
  warn: make('warn'),
  error: make('error'),
  fatal: make('fatal'),
};

const LOG_LEVEL_MAP: Record<string, string> = {
  'debug': 'debug',
  'info': 'info',
  'warn': 'warn',
  'warning': 'warn',
  'error': 'error',
  'fatal': 'fatal',
  'critical': 'fatal',
  'trace': 'debug',
};

export function setLogLevel(level: string): void {
  const normalized = LOG_LEVEL_MAP[level.toLowerCase()] ?? 'info';
  currentLevel = normalized;
}
